File: ROS/src/move_turtle/move_turtle/servo_sub.py
import time

import rclpy
import RPi.GPIO as gpio
from rclpy.node import Node
from std_msgs.msg import Float32


class ServoSub(Node):
    def __init__(self):
        super().__init__("servo_sub")
        # self.create_service(ServoAngle, "servo_angle", self.set_angle_callback)
        self.create_subscription(Float32, "servo_angle", self.set_angle_callback)
        self.servo_pin = 13
        gpio.setmode(gpio.BCM)
        gpio.setup(self.servo_pin, gpio.OUT)
        self.pwm = gpio.PWM(self.servo_pin, 50)
        self.pwm.start(0)
        self.theta = 0.0
        self.pwm.ChangeDutyCycle(2.5)
        time.sleep(1)
        self.pwm.ChangeDutyCycle(0)

    def set_angle_callback(self, msg: Float32):
        duty_cycle = 2.5 + (msg.data / 3.14 * 180 / 18.0)
        self.pwm.ChangeDutyCycle(duty_cycle)
        self.theta = msg.data

def main():
    rclpy.init()
    node = ServoSub()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        gpio.cleanup
        node.destroy_node()


if __name__ == "__main__":
    main()

File: ROS/src/move_turtle_cpp/src/servo_sub.cpp
#include "lifecycle_msgs/msg/transition_event.hpp"
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/float32.hpp"
#include <chrono>
#include <iostream>
#include <string>
#include <wiringPi.h>

using namespace std;
using namespace std::chrono_literals;

class ServoSub : public rclcpp::Node
{
public:
    ServoSub(const std::string &node_name, const std::string &pub_node)
        : Node(node_name), _servo_pin(13)
    {
        _sub = create_subscription<std_msgs::msg::Float32>(
            "servo_angle",
            10,
            std::bind(&ServoSub::sub_callback, this, std::placeholders::_1));
        _event_sub = create_subscription<lifecycle_msgs::msg::TransitionEvent>(
            pub_node + "/transition_event",
            10,
            std::bind(&ServoSub::event_sub_callback, this, std::placeholders::_1));
        wiringPiSetupGpio();
        pinMode(_servo_pin, PWM_OUTPUT);

        pwmSetMode(PWM_MODE_MS);
        pwmSetRange(1024);              // range for pwm
        pwmSetClock(375);       
    }

private:
    int _servo_pin = 13;
    rclcpp::Subscription<std_msgs::msg::Float32>::SharedPtr _sub;
    rclcpp::Subscription<lifecycle_msgs::msg::TransitionEvent>::SharedPtr _event_sub;
    void sub_callback(const std_msgs::msg::Float32::ConstSharedPtr msg)
    {
        int duty_cycle = 26 + (msg -> data * 1024) / 180;
        pwmWrite(_servo_pin, duty_cycle);
        RCLCPP_INFO(get_logger(), "data_callback : %f", msg->data);
    }
    void event_sub_callback(const lifecycle_msgs::msg::TransitionEvent::ConstSharedPtr msg)
    {
        RCLCPP_INFO(get_logger(), "start state : %s", msg->start_state.label.c_str());
        RCLCPP_INFO(get_logger(), "goal state : %s", msg->goal_state.label.c_str());
    }
};

int main(int argc, char *argv[])
{
    setvbuf(stdout, NULL, _IONBF, BUFSIZ);
    rclcpp::init(argc, argv);
    auto node = std::make_shared<ServoSub>("servo_litener", "servo_talker");
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}